### PR TITLE
runtime-rs: remove redundant sections in the config file

### PR DESCRIPTION
**Pull request by Tim-0731-Hzt**
1. we use annotation to choose which hypervisor and agent to use if multiple hypervisors and agents are configured in the same file

2. use the default hypervisor and agent if annotation is not set

Fixes: #5954
Signed-off-by: Zhongtao Hu <zhongtaohu.tim@linux.alibaba.com>

### CODE REVIEW

**Review by liubin — APPROVED** [review]

**Review by Ji-Xinyou — APPROVED** [review]

**Comment by c3d** [comment]
It looks like the code replaces an explicit default with "the first one". If so,

1. The commit message should change accordingly

2. The documentation probably needs updating.

That being said, having a named entry in a configuration file is self-explanatory, whereas picking up "the first one" is not. This is a serious drawback IMO. So what is the big benefit of this change that compensates for the drawback?

**Comment by Tim-0731-Hzt** [comment]
> It looks like the code replaces an explicit default with "the first one". If so,
> 
> 1. The commit message should change accordingly
> 2. The documentation probably needs updating.
> 
> That being said, having a named entry in a configuration file is self-explanatory, whereas picking up "the first one" is not. This is a serious drawback IMO. So what is the big benefit of this change that compensates for the drawback?

If 

> It looks like the code replaces an explicit default with "the first one". If so,
> 
> 1. The commit message should change accordingly
> 2. The documentation probably needs updating.
> 
> That being said, having a named entry in a configuration file is self-explanatory, whereas picking up "the first one" is not. This is a serious drawback IMO. So what is the big benefit of this change that compensates for the drawback?

Yes, I think so. But if multiple hypervisors and agents are set in one toml config file, how do we choose the default one?

**Comment by pmores** [comment]
So I believe that we want to support this feature (and I can see how it could be useful) we'll probably have to live with a config file syntax change.

To mitigate backwards compatibility issues, how about we reformulate the check so that when `runtime.hypervisor_name` is missing
- if `TomlConfig::hypervisor` contains exactly one item use that
- otherwise fail.

I believe this could fix up a vast majority (if not all) legacy cases since they are unlikely to have multiple hypervisors configured.

**Comment by Tim-0731-Hzt** [comment]
> So I believe that we want to support this feature (and I can see how it could be useful) we'll probably have to live with a config file syntax change.
> 
> To mitigate backwards compatibility issues, how about we reformulate the check so that when `runtime.hypervisor_name` is missing
> 
> * if `TomlConfig::hypervisor` contains exactly one item use that
> * otherwise fail.
> 
> I believe this could fix up a vast majority (if not all) legacy cases since they are unlikely to have multiple hypervisors configured.

I think maybe we can follow the go runtime, go runtime will choose the last item of the hashmap as the default one. https://github.com/kata-containers/kata-containers/blob/main/src/runtime/pkg/katautils/config.go#L1067

**Comment by pmores** [comment]
With arbitrary order, picking last is just as good as picking first since both are essentially equal to a random pick.

**Comment by Tim-0731-Hzt** [comment]
> With arbitrary order, picking last is just as good as picking first since both are essentially equal to a random pick.

Right @pmores, maybe we could keep the section `hypervisor_name` and `agent_name`, and let users choose the default. cc @jodh-intel

**Comment by YushuoEdge** [comment]
Or how about renaming these two sections to `default_hypervisor_name` and `default_agent_name` to be more straightforward?

**Comment by Tim-0731-Hzt** [comment]
@jodh-intel PTAL

File: src/libs/kata-types/src/config/runtime.rs [hunk 1]
```diff
@@ -17,14 +17,14 @@ pub const RUNTIME_NAME_VIRTCONTAINER: &str = "virt_container";
...
 
-    /// Hypervisor name: Plan to support dragonball, qemu
+    /// Hypervisor name: Plan to support dragonball, qemu, cloud hypervisor
     #[serde(default)]
     pub hypervisor_name: String,
 
-    /// Agent name
+    /// Agent name: Plan to support kata
     #[serde(default)]
     pub agent_name: String,
```
Comment (quanweiZhou) [inline src/libs/kata-types/src/config/runtime.rs:29]: I think `hypervisor_name` and `agent_name` need a default name.

File: src/runtime-rs/config/configuration-dragonball.toml.in [hunk 2]
```diff
@@ -207,8 +207,6 @@ container_pipe_size=@PIPESIZE@
 internetworking_model="@DEFNETWORKMODEL_DB@"
 
 name="@RUNTIMENAME@"
```
Comment (quanweiZhou) [inline src/runtime-rs/config/configuration-dragonball.toml.in:209]: According to your commit, I think the `runtime name` is also redundant.

File: src/runtime-rs/config/configuration-dragonball.toml.in [hunk 3]
```diff
@@ -207,8 +207,6 @@ container_pipe_size=@PIPESIZE@
...
-hypervisor_name="@HYPERVISOR_DB@"
-agent_name="@PROJECT_TYPE@"
```
Comment (jodh-intel) [inline src/runtime-rs/config/configuration-dragonball.toml.in:211]: Could you rebase and remove these options from the other runtime-rs config files?

File: src/libs/kata-types/src/config/runtime.rs (hunk 1 above)
Comment (jodh-intel) [reply src/libs/kata-types/src/config/runtime.rs:29]: Either that, or we need to return an error of either of these is the empty string.

File: src/runtime-rs/config/configuration-dragonball.toml.in (hunk 2 above)
Comment (jodh-intel) [reply src/runtime-rs/config/configuration-dragonball.toml.in:209]: Yes, please could you clarify this point. If users can configure different runtime types, they need to know what the types are and what the types mean.

The `runtime.name=` is not present in the golang config files so either we accept this is a rust-runtime specific option, or make it even clearer by changing the name of the table from `[runtime]` to `[runtime.@RUNTIMENAME@]`.

**Review by jodh-intel — COMMENTED** [review]
Thanks, @Tim-0731-Hzt. A few comments.

/cc @amshinde.

**Pull request closed without merging**